src/validators/CfnFnsSupportedFnsValidator.ts

```diff
@@ -17,26 +16,24 @@ const FN_NAME = '__FN_NAME__';
 export default class CfnFnsSupportedFnsValidator extends Validator {
   supportedFns: SupportedFns = {};
 
-  CfnFn(path: Path, propertyName: string, cfnFn: yaml.CfnFn) {
+  CfnFn(path: Path, _propertyName: string, cfnFn: yaml.CfnFn) {
     // Iterate over each segment of the path to determine if we're nested within
     // any functions that would not allow this ${cfnFn} to be used here
-    _.reduce(path, (currentPath: string[], p: string) => {
-
-      // We're looking at:
-      //  - [Root, SUPPORTED_FNS]
-      //  - [Root, Resources, SUPPORTED_FNS]
-      //  - [Root, Resources, A, SUPPORTED_FNS], etc
-      currentPath.push(p);
+    let currentPath: string[] = [...path];
+    while (currentPath.length) {
+      // Walk the path backwards looking for a parent CfnFn.
+      // If we find one that supports us, break.
+      // If we find one that doesn't, error.
+      currentPath.pop();
       const supportedFns = _.get(this.supportedFns, currentPath.concat(SUPPORTED_FNS));
-
-      if (_.isArray(supportedFns)) {
+      if (_.isArray(supportedFns)) { // We've hit a CfnFn node parent
         if (!_.includes(supportedFns, cfnFn.constructor)) {
           const fnName = _.get(this.supportedFns, currentPath.concat(FN_NAME))
           this.addError(path, `can not be used within ${fnName}`);
         }
+        break;
       }
-      return currentPath;
-    }, []);
+    }
 
     // Update supportedFns so that we can check any nested functions
     _.set(this.supportedFns, path.concat(SUPPORTED_FNS), cfnFn.supportedFns);
```

Comment:
This needs to go above the `pop` – otherwise this will get set on the wrong path.

Reply:
Never mind – `path` and `currentPath` (which is `pop`ed) are different arrays.